pkg/operator/ceohelpers/machine_api.go

```diff
@@ -63,3 +88,35 @@ func (m *MachineAPI) IsFunctional() (bool, error) {
 
 	return false, nil
 }
+
+// IsEnabled checks if the Machine API is enabled.
+// As of today Machine API is enabled when:
+// * clusterVersionInformer is synced
+// * clusterVersion has "MachineAPI" as an enabled capability
+func (m *MachineAPI) IsEnabled() (bool, error) {
+	if !m.hasVersionInformerSyncedFn() {
+		return false, fmt.Errorf("ClusterVersionInformer is not yet synced")
+	}
+
+	clusterVersion, err := m.versionLister.Get("version")
+	if err != nil {
+		return false, err
+	}
+
+	// this is a special case introduced from 4.14 on, where MachineAPI can be optional with clusters installed using capabilities.baselineCapabilitySet=None
+	if clusterVersion != nil && clusterVersion.Spec.Capabilities != nil && string(clusterVersion.Spec.Capabilities.BaselineCapabilitySet) == "None" {
+		machineAPIEnabled := false
+		for _, capability := range clusterVersion.Spec.Capabilities.AdditionalEnabledCapabilities {
```

Comment:
This won't be a reliable way to check, since any cluster that upgrades from a previous version prior to 4.14 won't have MachineAPI explicitly listed, but may have it enabled implicitly, this is recorded in status.

IMO, I would probably use the discovery client to check if the API is available within the cluster rather than looking at the configuration of the cluster version, it will be more reliable.

Something like this could be used
https://github.com/openshift/cluster-machine-approver/blob/702f688fb79b10e38f37ccc9cd61a1aac82de770/pkg/machinehandler/machinehandler.go#L108-L122

Reply:
Thanks for double-checking. I've taken the logic from the API check from https://github.com/openshift/api/pull/1537/files#diff-405d3ee50a80cb80e3a2a4f9b650043d55b14cfa350846d0dbb9f6d2e35dda08R440 :)

> upgrades from a previous version prior to 4.14 won't have MachineAPI explicitly listed

That's still what we assume, hence we're returning true at the bottom of the function (opt-in vs opt-out). Or will we have the case that an upgraded cluster has `BaselineCapabilitySet=None` and yet no mention of the MachineAPI (both in status+spec)?

> IMO, I would probably use the discovery client to check if the API is available within the cluster rather than looking at the configuration of the cluster version, it will be more reliable.

Agreed. I think we have the dynamic client sitting around somewhere for prometheus already. I'll try to replicate that using it.